### ISSUE TITLE
`multiprocessing` fails in MacOS due a `pickle` issue in a module in 10.5.0 [was: byexample fails to pickle an ExampleRunner]

### ISSUE THREAD

**Issue by saraedum** [byexamples/byexample#220]
In our CI setup, we are using a custom C++ runner to check the output of C++ code snippets.

When we upgrade byexample from 10.4 to 10.5, running byexample on our snippets fails with:
```
   [i:exec.cpp] Initializing C++ Interpreter
  [w:exec.cpp] Initialization of C++ Interpreter failed.
  [!] Something went wrong processing the file '/Users/runner/work/e-antic/e-antic/libeantic/test/byexample/../../../doc/manual/libeantic/overview_c.md':
  Traceback (most recent call last):
    File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/byexample/common.py", line 217, in human_exceptions
      yield o
    File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/byexample/byexample.py", line 22, in execute_examples
      return executor.execute(examples, filename)
    File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/byexample/log.py", line 267, in wrapped
      return func(*args, **kargs)
    File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/byexample/executor.py", line 271, in execute
      self.initialize_runners(runners, options)
    File "/usr/local/miniconda/envs/test/lib/python3.10/site-packages/byexample/executor.py", line 189, in initialize_runners
      runner.initialize(options)
    File "/Users/runner/work/e-antic/e-antic/libeantic/test/byexample/extensions/cxx.py", line 187, in initialize
      self._child.start()
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/process.py", line 121, in start
      self._popen = self._Popen(self)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/context.py", line 224, in _Popen
      return _default_context.get_context().Process._Popen(process_obj)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/context.py", line 284, in _Popen
      return Popen(process_obj)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
      super().__init__(process_obj)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
      self._launch(process_obj)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 47, in _launch
      reduction.dump(process_obj, fp)
    File "/usr/local/miniconda/envs/test/lib/python3.10/multiprocessing/reduction.py", line 60, in dump
      ForkingPickler(file, protocol).dump(obj)
  _pickle.PicklingError: Can't pickle <class 'cxx.CxxInterpreter'>: it's not the same object as cxx.CxxInterpreter
```

The problem only appears in CI runs on macOS (works fine on Ubuntu in CI and also locally on ArchLinux.) It seems to be independent of the Python version, since downgrading to Python 3.9 did not fix this.

I have not tried to run this on a macOS machine myself yet, so I have no better reproducer to offer at the moment, than our full CI setup: it calls byexample through automake here https://github.com/flatsurf/e-antic/blob/master/.github/workflows/test.yml#L103

Obviously this is not a great help to debug it. I'll try to understand better myself what is going on here but maybe somebody has an idea already what might be the problem?

The implementation of our custom runner is here: https://github.com/flatsurf/e-antic/blob/master/libeantic/test/byexample/extensions/cxx.py

**Expected behavior**
We expected byexample to run our tests without an error instead of raising an error.

**byexample version**
We are seeing this problem with byexample 10.5.0 (build `py310h2ec42d9_0`) and python 3.10.2 (build `hea1dfa3_3_cpython`) as distributed by conda-forge.

We don't see the issue when downgrading byexample to 10.4.0 (build `py310h2ec42d9_1`.)

**Comment by eldipa** [comment]
Thanks for the report!

Quick two possible workarounds:

 - use a previous version of `byexample`, the offending [changeset](https://github.com/byexamples/byexample/commit/9f2a0fee45e68447265a8339a816773c3cf2e036) was introduced in 10.5.0 so any previous version should work. The only catch is that you will have to use Python 3.9.
 - at the begin of `initialize` and these lines:

```python
def initialize(self, options):
        import multiprocessing
        multiprocessing.set_start_method('fork')
        ...
```

Let me know if one of those workarounds work for you.

By the way, if you replace `'fork'` with `'spawn'` you should be able to reproduce the bug in any linux distribution like Ubuntu.

## Bug analysis

What it is happening is that in MacOS `multiprocessing` uses the `"spawn"` method which requires the serialization (pickle) of all the objects that are shared between the parent and the child processes. In this case, the class `CxxInterpreter` and the pipe.

The class *can* be pickled without problem, it is not using any attribute that couldn't be pickled.

The parent and the child processes require to use the **same** class definition (and in general, the same code definition) and here is where we find the issue.

The parent and the child disagree, each have a "different" definition of `CxxInterpreter` (hence the confusing error `Can't pickle <class 'cxx.CxxInterpreter'>: it's not the same object as cxx.CxxInterpreter`).

Why is happening now? `byexample` version 10.5.0 modified how the modules/extensions are loaded. Previous 10.5.0, `byexample` used  libs/methods that are deprecated since Python 3.3 (I think) and were removed in Python 3.10.

So in 10.5.0 I changed the code to support Python 3.10 (this is the [changeset](https://github.com/byexamples/byexample/commit/9f2a0fee45e68447265a8339a816773c3cf2e036) )

For some reason, this way to load the modules/extensions is incompatible with pickle.

I'll investigate a little more in this weekend.

**Comment by eldipa** [comment]
### Minimum reproducible example

Assuming a folder named `plugins/` with a python file with a class named `Foo` like the following:

```python
class Foo:
    pass
```

The following `load_plugin_good.py` loads and pickles correctly the class `Foo`. This is what `byexample` pre-10.5.0 was doing but it is not supported by Python 3.10:

```python
import sys, pkgutil, inspect, pickle
import importlib.util

dirnames = ["plugins/"]
objs = []

# For each plugin folder, see which Python files are there
# and load them
for importer, name, is_pkg in pkgutil.iter_modules(dirnames):

    # Find and load the Python module
    module = importer.find_module(name).load_module(name)

    # Only load the classes from the module
    classes = inspect.getmembers(module, lambda obj: inspect.isclass(obj))

    # Create one object per class
    for _, class_ in classes:
        objs.append(class_())

# Pickle the object
obj = objs[0]
print(pickle.dumps(obj))
```

The following `load_plugin_bad.py` is what is currently implemented in `byexample` 10.5.0 and reproduces the issue:

```python
import sys, pkgutil, inspect, pickle
import importlib.util

dirnames = ["plugins/"]
objs = []

# For each plugin folder, see which Python files are there
# and load them
for importer, name, is_pkg in pkgutil.iter_modules(dirnames):

    # Find and load the Python module
    spec = importer.find_spec(name)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    # Only load the classes from the module
    classes = inspect.getmembers(module, lambda obj: inspect.isclass(obj))

    # Create one object per class
    for _, class_ in classes:
        objs.append(class_())

# Pickle the object
obj = objs[0]
print(pickle.dumps(obj))
```

The traceback obtained is:

```python
Traceback (most recent call last):
  File "/home/user/tmp/byexample-pickle/load_plugin_bad.py", line 25, in <module>
    print(pickle.dumps(obj))
_pickle.PicklingError: Can't pickle <class 'foo.Foo'>: import of module 'foo' failed
```

Notice that pickle cannot import the module `foo`. This can be *workaround'ed* adding `plugins/` to the `sys.path` however that would be hiding the root issue.

**Comment by eldipa** [comment]
### Minimum reproducible example (pickle + multiprocessing)

Assuming a folder named `plugins/` with a python file with a class named `Foo` like the following:

```python
class Foo:
    @classmethod
    def sayhi(cls):
        print("Hi!")
```

The following `load_plugin_bad.py` is what is currently implemented in `byexample` 10.5.0 and reproduces the issue:

```python
import sys, pkgutil, inspect, pickle
import importlib.util
import multiprocessing

def init():
    dirnames = ["plugins/"]
    objs = []

    # For each plugin folder, see which Python files are there
    # and load them
    for importer, name, is_pkg in pkgutil.iter_modules(dirnames):

        # Find and load the Python module
        spec = importer.find_spec(name)
        module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(module)

        # XXX Fix 1
        sys.modules[name] = module

        # Only load the classes from the module
        classes = inspect.getmembers(module, lambda obj: inspect.isclass(obj))

        # Create one object per class
        for _, class_ in classes:
            print(class_.__module__)
            objs.append(class_())

    return objs

def try_spawn(obj):
    # Pickle the object
    print(pickle.loads(pickle.dumps(obj)))

    ch = multiprocessing.Process(target=obj.sayhi)
    ch.start()
    ch.join()

if __name__ == '__main__':
    multiprocessing.set_start_method('spawn')
    objs = init()
    try_spawn(objs[0])
```

With the `Fix 1`, we are adding the loaded module into `sys.modules`. This used to happen by default but it is not longer the default with the new import mechanism in `byexample` 10.5.0.

That fixes the `pickle` issue however it is not the end of the story. `multiprocessing` is able to pickle `obj` but it is not able to load it.

This is the traceback:

```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/user/env/Python-3.9.7/Lib/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/home/user/env/Python-3.9.7/Lib/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
ModuleNotFoundError: No module named 'foo'
```

Basically the `spawn` method uses a separated Python interpreter, it does **not** load the modules that `byexample` has and therefore it fails to load `foo`. This is known limitation of `pickle`.

I will check if there is a solution for this or if there is another "plugin" library implemented out there with better support.

@saraedum when you have time, let me know if one of the workarounds that I wrote in the first post worked.

The real fix is going to take more time that I was expecting.

**Comment by saraedum** [comment]
> when you have time, let me know if one of the workarounds that I wrote in the first post worked.

Sure, I am just using a previous version of byexample. I don't need to run byexample tests with Python 3.10.
